File: src/runningbacks.jl
using Query, Statistics
include("query.jl")
include("utility.jl")

DB = LibPQ.Connection("dbname=nfldb")
MIN_YEAR = 2018

# load key datasets
rb = query(DB, read("queries/runningbacks.sql", String))
balance = query(DB, "select * from team_balance")

# tendency of team to use run plays
runchance = balance |>
    @groupby({_.offense, _.year}) |>
    @map({team=key(_).offense, year=key(_).year, pct_pass=mean(_.pct_pass), pct_run=mean(_.pct_run)}) |>
    @orderby_descending(_.pct_run) |>
    DataFrame

# chances of getting the carry on a run play
runstats = rb |>
    @groupby({_.team, _.year, _.name}) |>
    @map({
        year=key(_).year,
        team=key(_).team,
        player=key(_).name,
        rushes=sum(_.rushes),
        yds=sum(_.runyds),
        avg=mean(_.runyds),
        #vol=std(_.runyds)
    }) |>
    @orderby({_.year, _.team}) |>
    DataFrame

teamruns = runstats |>
    @groupby({_.team, _.year}) |>
    @map({team=key(_).team, year=key(_).year, rushes=sum(_.rushes)}) |>
    DataFrame
runshare = runstats |>
    @join(teamruns, {_.team, _.year}, {_.team, _.year}, {_.team, _.year, _.player, runshare=_.rushes/__.rushes}) |>
    DataFrame
runstats = runstats |>
    @join(runshare, {_.team, _.year, _.player}, {_.team, _.year, _.player}, {_.team, _.year, _.player, _.rushes, __.runshare, _.yds, _.avg,}) |>
    @join(runchance, {_.team, _.year}, {_.team, _.year}, {_.team, _.year, _.player, runchance=__.pct_run, _.runshare, _.rushes, _.yds, _.avg}) |>
    DataFrame

runstats |> @filter(_.year>=MIN_YEAR) |> @orderby_descending(_.runchance*_.runshare*_.avg) |> DataFrame

File: src/runningbacks.r
library(data.table)
library(RPostgreSQL)

RunQuery = function(connection, query){
  result = dbSendQuery(connection, query)
  out = fetch(result, n=-1)
  dbClearResult(result)
  return(as.data.table(out))
}

if (!exists("CONNECTION")){
  CONNECTION = dbConnect(PostgreSQL(), user="nfldb", dbname="nfldb", host="localhost")
}

query = paste(readLines("queries/runningbacks.sql"), collapse=' ')
rb_plays = RunQuery(CONNECTION, query)
rb_plays[,yardline:=as.integer(gsub('[()]','',yardline))+50]

rb = rb_plays[year==2018]

X = rb[,
       .(down=mean(down,na.rm=T),
         spot=mean(yardline),
         runs=sum(rushes),
         tgts=sum(targets),
         touches=sum(rushes)+sum(targets),
         yds=sum(runyds)+sum(recyds),
         vol=sd(runyds+recyds),
         relyds=mean((runyds+recyds)/togo,na.rm=T)),
       by=.(team,name)]
X[, score:=touches*yds*relyds*spot/vol]
X = X[touches>median(touches)][order(score)]

par(cex=0.5, family='mono', mar=c(2.5,2.5,1.5,0.5), mgp=c(1.5,0.5,0))
b = barplot(X[,score], horiz=T)
text(x=X[,score], y=b[,1], labels=X[,name])
box(bty='l')
grid(lty=1, nx=NULL, ny=0, col='#00000020')
